mk-python-derivation: add pythonRelaxDepsHook

mweinelt opened the pull request:
Configuring `pythonRelaxDeps` or `pythonRemoveDeps` does not require
adding `pythonRelaxDepsHook` into `nativeBuildInputs` anymore.

Simple alternative to #319432

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
mk-python-derivation: add pythonRelaxDepsHook

Configuring `pythonRelaxDeps` or `pythonRemoveDeps` does not require
adding `pythonRelaxDepsHook` into `nativeBuildInputs` anymore.

Co-Authored-By: superherointj <5861043+superherointj@users.noreply.github.com>

superherointj commented:
We are doing duplicated work. I'd have preferred your review on the original PR. Now we have two PRs.
Also, I think, an empty list, false boolean, or invalid types should not trigger the hook. Doesn't look beautiful but has a purpose to it.
Anyway, to solve the PR conflict, I'm closing my PR.

mweinelt commented:
I'm sorry, I lacked the spoons to jump into that conversation. I primarily created because I felt that your approach was overcomplicating the idea. I wanted to try out this simpler approach and did the treewide change (albeit only for python-modules), to test its effect. By the time I had done that, it was working so well, that I just commited and pushed it.

thiagokokada commented:
> We are doing duplicated work. I'd have preferred your review on the original PR. Now we have two PRs.
> Also, I think, an empty list, false boolean, or invalid types should not trigger the hook. Doesn't look beautiful but has a purpose to it.
> Anyway, to solve the PR conflict, I'm closing my PR.

The hook itself shouldn't need that much baby sitting, it will only trigger if the value is true or is a list.

mweinelt commented:
It also triggers for a string, e.g. "foo bar baz" and an integer like `1`, which seems to map to true? Either way, not problematic.

mweinelt committed:
docs/language-frameworks/python: update relaxDepsHook docs

phaer commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -252,6 +252,7 @@ let
       #    because the hook that checks for conflicts uses setuptools.
       #
       pythonCatchConflictsHook
+    ] ++ optionals (attrs ? pythonRelaxDeps || attrs ? pythonRemoveDeps) [
```
You are just adding an empty list here? Looks like the tree-wide change in the last commit might have mistakenly deleted pythonRelaxDepsHook here? :D

mweinelt replied:
Thanks for the review!

phaer reviewed (changes requested):
Thanks! I think this is a good idea! :)

mweinelt committed:
treewide: remove pythonRelaxDepsHook references

It is is now provided automatically, when `pythonRelaxDeps` or
`pythonRemoveDeps` is defined through `mk-python-derivation`.

SuperSandro2000 reviewed (approved):
LGTM
